File: installers/v2/windows/tray-app/UpdateChecker.cs
using System;

namespace Tadaima.Tray;

/// <summary>
/// "Check for Updates" — query npm for the latest
/// @psychout98/tadaima version, compare to the installed version, and
/// optionally apply the update by stopping the scheduled task, running
/// npm install, and restarting the task.
/// </summary>
internal static class UpdateChecker
{
    public readonly record struct UpdateInfo(string Current, string Latest);

    public static UpdateInfo? Check()
    {
        var npmResult = Shell.Run(BundlePaths.NodeExe, new[]
        {
            BundlePaths.NpmCliJs,
            "view",
            "@psychout98/tadaima",
            "version",
        });
        if (npmResult.ExitCode != 0)
        {
            throw new InvalidOperationException(
                "npm view failed: " +
                (string.IsNullOrWhiteSpace(npmResult.StdErr) ? npmResult.StdOut : npmResult.StdErr));
        }

        var latest = npmResult.StdOut.Trim();
        var local = Shell.Run(BundlePaths.NodeExe, new[]
        {
            System.IO.Path.Combine(BundlePaths.AgentPrefix, "node_modules", "@psychout98", "tadaima", "dist", "index.js"),
            "--version",
        });
        if (local.ExitCode != 0)
        {
            throw new InvalidOperationException("tadaima --version failed: " + local.StdErr);
        }

        var current = ParseVersion(local.StdOut);
        if (CompareSemver(current, latest) < 0)
        {
            return new UpdateInfo(current, latest);
        }
        return null;
    }

    public static void Apply()
    {
        // 1. Stop the scheduled task.
        Shell.Run("schtasks.exe", new[] { "/End", "/TN", "Tadaima Agent" });

        // 2. npm install -g --prefix <install>\agent @psychout98/tadaima@latest
        var install = Shell.Run(BundlePaths.NodeExe, new[]
        {
            BundlePaths.NpmCliJs,
            "install",
            "-g",
            "--prefix",
            BundlePaths.AgentPrefix,
            "@psychout98/tadaima@latest",
        });
        if (install.ExitCode != 0)
        {
            throw new InvalidOperationException("npm install failed: " + install.StdErr);
        }

        // 3. Restart the scheduled task.
        Shell.Run("schtasks.exe", new[] { "/Run", "/TN", "Tadaima Agent" });
    }

    private static string ParseVersion(string cliOutput)
    {
        var s = cliOutput.Trim();
        var lastSpace = s.LastIndexOf(' ');
        if (lastSpace >= 0 && lastSpace + 1 < s.Length) s = s[(lastSpace + 1)..];
        return s.StartsWith('v') ? s[1..] : s;
    }

    // Very small semver comparator — good enough for "current < latest".
    private static int CompareSemver(string a, string b)
    {
        var (ac, ap) = Split(a);
        var (bc, bp) = Split(b);
        var max = Math.Max(ac.Length, bc.Length);
        for (var i = 0; i < max; i++)
        {
            var ai = i < ac.Length ? ac[i] : 0;
            var bi = i < bc.Length ? bc[i] : 0;
            if (ai != bi) return ai < bi ? -1 : 1;
        }
        if (ap == bp) return 0;
        if (string.IsNullOrEmpty(ap)) return 1;
        if (string.IsNullOrEmpty(bp)) return -1;
        return string.CompareOrdinal(ap, bp);
    }

    private static (int[] core, string pre) Split(string v)
    {
        var parts = v.Split('-', 2);
        var core = Array.ConvertAll(parts[0].Split('.'), int.Parse);
        var pre = parts.Length > 1 ? parts[1] : "";
        return (core, pre);
    }
}

File: installers/v2/macos/menubar-app/Sources/TadaimaMenuBar/UpdateChecker.swift
import Foundation

/// "Check for Updates" action: queries npm for the latest
/// @psychout98/tadaima version, compares it to the installed version,
/// and if newer offers a one-click update.
enum UpdateChecker {
    struct UpdateInfo {
        let current: String
        let latest: String
    }

    enum CheckError: Error, LocalizedError {
        case npmFailed(String)
        case localVersionFailed(String)
        case parseFailed(String)

        var errorDescription: String? {
            switch self {
            case .npmFailed(let msg): return "npm view failed: \(msg)"
            case .localVersionFailed(let msg): return "tadaima --version failed: \(msg)"
            case .parseFailed(let msg): return "could not parse version: \(msg)"
            }
        }
    }

    static func check() throws -> UpdateInfo? {
        // Ask npm for the latest version. We spawn the bundled Node
        // against the bundled npm-cli.js to avoid depending on a system
        // Node that may not exist.
        let npmResult = Shell.run(
            BundlePaths.nodeBinary,
            [
                BundlePaths.npmCliJs.path,
                "view",
                "@psychout98/tadaima",
                "version",
            ]
        )
        guard npmResult.exitCode == 0 else {
            throw CheckError.npmFailed(npmResult.stderr.isEmpty ? npmResult.stdout : npmResult.stderr)
        }
        let latest = npmResult.stdout.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !latest.isEmpty else {
            throw CheckError.parseFailed("empty latest version")
        }

        // Local version: spawn the installed tadaima CLI via the bundled
        // Node. `tadaima --version` prints "tadaima-agent vX.Y.Z", we
        // parse the last whitespace-separated token.
        let localResult = Shell.run(
            BundlePaths.nodeBinary,
            [BundlePaths.agentBinary.path, "--version"]
        )
        guard localResult.exitCode == 0 else {
            throw CheckError.localVersionFailed(
                localResult.stderr.isEmpty ? localResult.stdout : localResult.stderr
            )
        }
        let tokens = localResult.stdout
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .split(separator: " ")
        guard let versionToken = tokens.last else {
            throw CheckError.parseFailed("no version token in: \(localResult.stdout)")
        }
        let current = String(versionToken).replacingOccurrences(of: "v", with: "")

        if compareSemver(current, latest) < 0 {
            return UpdateInfo(current: current, latest: latest)
        }
        return nil
    }

    static func applyUpdate() throws {
        // 1. Stop the agent via launchctl.
        _ = Shell.run(
            URL(fileURLWithPath: "/bin/launchctl"),
            ["bootout", "gui/\(Shell.currentUid())", BundlePaths.launchAgentPlist.path]
        )
        // 2. npm install -g --prefix <agent-prefix> @psychout98/tadaima@latest
        let install = Shell.run(
            BundlePaths.nodeBinary,
            [
                BundlePaths.npmCliJs.path,
                "install",
                "-g",
                "--prefix",
                BundlePaths.agentPrefix.path,
                "@psychout98/tadaima@latest",
            ]
        )
        if install.exitCode != 0 {
            throw CheckError.npmFailed(install.stderr)
        }
        // 3. Bring the agent back up.
        _ = Shell.run(
            URL(fileURLWithPath: "/bin/launchctl"),
            ["bootstrap", "gui/\(Shell.currentUid())", BundlePaths.launchAgentPlist.path]
        )
    }

    /// Very small semver comparator — we only need to decide
    /// "current < latest". Pre-release tags are treated as < the same
    /// base version, which matches npm's behavior well enough for the
    /// tray update flow.
    static func compareSemver(_ a: String, _ b: String) -> Int {
        let (ac, ap) = split(a)
        let (bc, bp) = split(b)
        let maxLen = max(ac.count, bc.count)
        for i in 0..<maxLen {
            let ai = i < ac.count ? ac[i] : 0
            let bi = i < bc.count ? bc[i] : 0
            if ai != bi { return ai < bi ? -1 : 1 }
        }
        if ap == bp { return 0 }
        if ap.isEmpty { return 1 }   // a is release, b is prerelease → a > b
        if bp.isEmpty { return -1 }  // b is release, a is prerelease → a < b
        return ap < bp ? -1 : 1
    }

    private static func split(_ v: String) -> ([Int], String) {
        let parts = v.split(separator: "-", maxSplits: 1, omittingEmptySubsequences: false)
        let core = String(parts[0]).split(separator: ".").compactMap { Int($0) }
        let pre = parts.count > 1 ? String(parts[1]) : ""
        return (core, pre)
    }
}
